### PR TITLE
`azurerm_firewall_policy` - support for `explicit_proxy` and `auto_learn_private_ranges_mode` properties

### PR DESCRIPTION
**Pull request by wuxu92**
test result:

```
=== RUN   TestAccFirewallPolicy_complete
=== PAUSE TestAccFirewallPolicy_complete
=== CONT  TestAccFirewallPolicy_complete
--- PASS: TestAccFirewallPolicy_complete (145.61s)
PASS

```

### CODE REVIEW

File: internal/services/firewall/firewall_policy_resource.go [hunk 1]
```diff
@@ -925,6 +1020,15 @@ func resourceFirewallPolicySchema() map[string]*pluginsdk.Schema {
...
 
+		"auto_learn_private_ranges_mode": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+			ValidateFunc: validation.StringInSlice([]string{
+				string(network.AutoLearnPrivateRangesModeEnabled),
+				string(network.AutoLearnPrivateRangesModeDisabled),
+			}, false),
+		},
+
```
Comment (catriona-m) [inline internal/services/firewall/firewall_policy_resource.go:1031]: Would it make sense to make this property a boolean and call it `auto_learn_private_ranges_enabled` instead?

Comment (wuxu92) [reply internal/services/firewall/firewall_policy_resource.go:1031]: that makes sense and more easy to use, but I wonder if there are other possible values in future

Comment (catriona-m) [reply internal/services/firewall/firewall_policy_resource.go:1031]: Typically in cases where there is only enabled and disabled we tend to use a boolean anyway as it's not likely to have different values in future in this case.

Comment (wuxu92) [reply internal/services/firewall/firewall_policy_resource.go]: thanks, modified.

**Review by catriona-m — CHANGES_REQUESTED** [review]
Thanks @wuxu92, this is looking pretty good now, although it seems like there are some test failures that may need fixing:

```

Test ended in panic.
------- Stdout: -------
=== RUN   TestAccFirewallPolicy_basic
=== PAUSE TestAccFirewallPolicy_basic
=== CONT  TestAccFirewallPolicy_basic
------- Stderr: -------
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x44e566b]
goroutine 420 [running]:
github.com/hashicorp/terraform-provider-azurerm/internal/services/firewall.resourceFirewallPolicyRead(0xc003063600, {0x5bc00e0?, 0xc003734000?})
  /opt/teamcity-agent/work/5e6516bb4d10eb66/internal/services/firewall/firewall_policy_resource.go:213 +0x11ab
github.com/hashicorp/terraform-provider-azurerm/internal/services/firewall.resourceFirewallPolicyCreateUpdate(0xc003063600, {0x5bc00e0?, 0xc003734000})
  /opt/teamcity-agent/work/5e6516bb4d10eb66/internal/services/firewall/firewall_policy_resource.go:137 +0x15c5
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x6e29880?, {0x6e29880?, 0xc0017ce7b0?}, 0xd?, {0x5bc00e0?, 0xc003734000?})
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:695 +0x178
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc00398eb60, {0x6e29880, 0xc0017ce7b0}, 0xc0027a89c0, 0xc003063480, {0x5bc00e0, 0xc003734000})
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:837 +0xa85
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc0032715a8, {0x6e29880?, 0xc0017ce6f0?}, 0xc001e3a9b0)
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/grpc_provider.go:1021 +0xe8d
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc001b80780, {0x6e29880?, 0xc001828e40?}, 0xc003056e70)
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server/server.go:818 +0x574
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x623d800?, 0xc001b80780}, {0x6e29880, 0xc001828e40}, 0xc003056e00, 0x0)
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001ec000, {0x6e3d3e0, 0xc0025541a0}, 0xc001b10d80, 0xc00183c450, 0xb179fa0, 0x0)
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/google.golang.org/grpc/server.go:1340 +0xd23
google.golang.org/grpc.(*Server).handleStream(0xc0001ec000, {0x6e3d3e0, 0xc0025541a0}, 0xc001b10d80, 0x0)
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/google.golang.org/grpc/server.go:1713 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/google.golang.org/grpc/server.go:965 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/google.golang.org/grpc/server.go:963 +0x28a
```

**Comment by wuxu92** [comment]
@catriona-m all test cases pass now

image

**Review by katbyte — APPROVED** [review]
LGTM 🍏

**Pull request merged**